sage.py

```diff
@@ -571,393 +485,271 @@ def getepisodes(action_seq, mcat, plot, debug=False):
-                            rest = (end+1, len(ep)-1)
-                            al = episodes[start:end+1]
-                            if len(al) < 2:
-                                #print('discarding symbol ', [x[2] for x in al])#, start, end, len(episodes))
-                                continue
-                            att = atta+'->'+victim+'-'+str(c)
-                            #print('---', att, [x[2] for x in al])
-                            keys.append(att)
-                            alerts.append(al)
-                            c+=1
-                        #print('--', ep[rest[0]: rest[1]+1])
-                        al = episodes[rest[0]: rest[1]+1]
-                        if len(al) < 2:
-                            #print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
-                            continue
-                        att = atta+'->'+victim+'-'+str(c)
-                        #print('---', att, [x[2] for x in al])
-                        keys.append(att)
-                        alerts.append(al)
-    print('\n# sub-sequences', len(keys))
-    return (alerts, keys)
-
-## 27 Aug 2020: Generating traces for flexfringe 
-def generate_traces(alerts, keys, datafile, test_ratio=0.0):
-    victims = alerts
-    al_services = [[most_frequent(y[6]) for y in x] for x in victims]
-    print('----- All unique servcies -----')
-    print(set([item for sublist in al_services for item in sublist]))
+            # Cut episode sequence when a low-severity episode follows a high-severity episode
+            count = 0
+            mcats = [epi[2] for epi in episodes]
+            cuts = [i for i in range(len(episodes) - 1) if (len(str(mcats[i])) > len(str(mcats[i + 1])))]#(ep[i] > 100 and ep[i+1] < 10)]
+            if len(cuts) == 0:
+                attacks.append(attack + '-0')
+                subsequences.append(episodes)
+                continue
+
+            rest = (-1, -1)
+            for i in range(len(cuts)):
+                start = 0 if i == 0 else cuts[i - 1] + 1
+                end = cuts[i]
+                rest = (end + 1, len(episodes) - 1)
+                subsequence = episodes[start:end+1]
+                if len(subsequence) < 2:
+                    continue
+                attacks.append(attack + '-' + str(count))
+                subsequences.append(subsequence)
+                count += 1
+            subsequence = episodes[rest[0]:rest[1]+1]
+            if len(subsequence) < 2:
+                #print('discarding symbol ', [x[2] for x in al]) # TODO This one is not cool1
+                continue
+            attacks.append(attack + '-' + str(count))
+            subsequences.append(subsequence)
+    print('\n# sub-sequences', len(attacks))
+    return subsequences, attacks
+
+
+## 27 Aug 2020: Generating traces for FlexFringe
+def generate_traces(subsequences, datafile):
+    all_services = [[most_frequent(epi[6]) for epi in subseq] for subseq in subsequences]
```

Comment:
What is the purpose of these lines? Is it to show all the unique services that occur in the AGs (so only mserv's should be counted), or to show all unique services present in all episodes?

In the latter case, `most_frequent` call will discard some episodes and the count of services will be lower than actually (which confused me a bit during RP). In the former case, the log string is misleading; also, these services can be counted in the for-loop below so that they are not recomputed twice.

Reply:
It was used when we were grouping services by themes. Sometimes, new services would show up and since the themes dictionary was manually created, there would be errors. This can be removed now that we directly show the service itself.

Reply:
Just to be sure, do you mean lines 655-658 in the screenshot below?

![image](https://github.com/tudelft-cda-lab/SAGE/assets/45541174/4ee8500c-abd1-4468-8c10-00f29f81a407)

Reply:
[Resolved during the meeting]